use a file based startup on darwin

noamokman opened the pull request:
Hi there!
Should close #25 #28 #31 

Thanks!

noamokman commented:
@adam-lynch would love your opinion :smile:

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,58 @@
+            file = @getFile(opts)
+            array = [opts.appPath]
+            if(opts.isHiddenOnLaunch) then array.push('--hidden')
+            command = array.map(a => `    <string>${a}</string>`)
```
Wait... backticks in CoffeeScript? Never saw this before. I thought you had to use double quotes for interpolation.

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,58 @@
+                .join('\n')
+
+            data = [
+                '<?xml version="1.0" encoding="UTF-8"?>',
```
CoffeeScript has support for multiline strings. See http://coffeescript.org/#strings

adam-lynch commented:
Thanks. Educate me... what are the pros & cons of this versus the existing way?

If we did change it, what about app's that were using the old version? How do they transition to this? Do we need to keep the old logic in this module and transition it to file-based?

noamokman commented:
Thanks for the remarks!
Fixed the problematic bits.
Also removed the applescript dep.

I think this should be a breaking change,
so we wont have to deal with mixed implementation.

As for pros and cons,
It seems that the current way has problems with a terminal window popping up.
I believe this will work better

P.S,
This unifies the `--hidden` argument being sent on all platforms,
so it's already a good enough reason to bump major.
Maybe write more clearly that the hidden option sends this `--hidden` arg on startup,
and that is up for the developer to use relevant logic based on the args.

dcposch commented:
> It seems that the current way has problems with a terminal window popping up.

Are there projects that support Mac and use `auto-launch` in production?

I'm surprised this hasn't come up before--it seems like creating a terminal window in addition to your app after every boot would result in lots of user feedback & bug reports. (Esp since closing the terminal window then kills the app.)

adam-lynch commented:
@dcposch I can say yes definitely with NW.js but can't confirm Electron but I assume so?

@noamokman I'll take a proper look at this soon. In the meantime, maybe others could confirm this is OK for them / solves their problems? I'll ask in the issues you've linked

mzdr commented:
Hey @noamokman, your patch gives me the following error:

```js
Promise {
  <rejected> ReferenceError: name is not defined
    at /Users/sp/Projects/timestamp/node_modules/auto-launch/dist/AutoLaunchMac.js:29:236
    at Object.enable (/Users/sp/Projects/timestamp/node_modules/auto-launch/dist/AutoLaunchMac.js:18:12)
    at AutoLaunch.module.exports.AutoLaunch.enable (/Users/sp/Projects/timestamp/node_modules/auto-launch/dist/index.js:72:21)
    at new App (/Users/sp/Projects/timestamp/app/app.js:45:39)
    at App.Electron.app.on (/Users/sp/Projects/timestamp/app/index.js:7:32)
    at emitOne (events.js:101:20)
    at App.emit (events.js:188:7) }
```

Have a look at https://github.com/Teamwork/node-auto-launch/pull/38/files#diff-3cb4ac24dd2f84b466e7af980f600036R25. Or am I wrong?

mzdr commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,56 @@
+                    <plist version="1.0">
+                    <dict>
+                      <key>Label</key>
+                      <string>#{name}</string>
```
Getting `ReferenceError: name is not defined` here…

noamokman commented:
@mzdr oops, please try again,
Thanks!

mzdr commented:
Out of the box it's not working on my machine (Electron v1.3.4 and OSX 10.11.6). There is no `.plist` file being written and I don't get any errors. I am debugging right now…

mzdr commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,56 @@
 
 module.exports =
+    getFile: (opts) ->
+        file = dir+opts.appName+'.plist'
```
@noamokman I think there is a `/` missing here because I keep getting file paths like this: `/Users/sp/Library/LaunchAgentsMinecraft.plist`

noamokman replied:
great catch! thanks for the help :smile:

noamokman commented:
@mzdr added the missing  `/`

mzdr commented:
Okay, now it seems to be working. After logging out and back in again my app was automatically started. 👍 

Though I personally have some concerns with the LaunchAgent approach. For example if you delete an app which was set to auto launch, the `.plist` file will end up as a dead file on the users hard disk. I know it’s picky, but nevertheless it’s something we should care about. 

Furthermore the average (or more sophisticated) user will probably check his login items under `System Preferences → Users & Groups → Login Items` and delete old/non-existent apps from there. But apps registered as a LaunchAgent won’t show up there.

@adam-lynch I think this is something we should consider.

noamokman commented:
> Though I personally have some concerns with the LaunchAgent approach. For example if you delete an app which was set to auto launch, the .plist file will end up as a dead file on the users hard disk. I know it’s picky, but nevertheless it’s something we should care about.

@mzdr
This is also true for the [linux implementation](https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchLinux.coffee).

dcposch commented:
@mzdr @adam-lynch Thoughts?

For WebTorrent Desktop, we can't use auto-launch as long as it's creating a terminal window on Mac.

mzdr commented:
@dcposch Did you checkout the workaround posted in [#28](https://github.com/Teamwork/node-auto-launch/issues/28#issuecomment-234425664)? As for now I'm using auto-launch with the workaround as well…

dcposch commented:
@mzdr yeah, I tried the workaround for Electron listed here:

https://github.com/Teamwork/node-auto-launch/issues/28#issuecomment-222194437

Didn't work

adam-lynch commented:
OK I'm going to test this tonight/tomorrow. I'll give it a quick review now.

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
-            applescript.execString command, (err) ->
+        new Promise (resolve, reject) =>
+            file = @getFile(opts)
+            array = [opts.appPath]
```
Please don't call this `array`

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
+        new Promise (resolve, reject) =>
+            file = @getFile(opts)
+            array = [opts.appPath]
+            if(opts.isHiddenOnLaunch) then array.push('--hidden')
```
You don't need the `then`; you can do this:

```coffeescript
array.push('--hidden') if opts.isHiddenOnLaunch
```

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
+                    </dict>
+                    </plist>"""
+
+            mkdirp.sync(dir)
```
Shouldn't we have some error checking here?

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
     disable: (opts) ->
-        new Promise (resolve, reject) ->
-            command = tellTo + "delete login item \"#{opts.appName}\""
+        new Promise (resolve, reject) =>
```
Since you're touching these lines (where `new Promise` is returned), do you mind adding a `return` at the start? I'm not a fan of using implicit `return`s in multi-line functions

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
+            fs.stat file, (err) ->
                 return reject(err) if err?
-                resolve()
+                fs.unlink file, (err2) ->
```
This is silly but name it something other than `err2`. Even if it's `error` or something.

adam-lynch commented on src/AutoLaunchMac.coffee:
```diff
@@ -1,33 +1,55 @@
-                isPresent = loginItems?.indexOf(opts.appName)
-                resolve(isPresent? and isPresent isnt -1)
+            fs.stat file, (err, stat) ->
+    # TODO: Error handling
```
Need to reject here.

adam-lynch commented:
One other thing I noticed is that the app doesn't appear in the user's Login Items (in system preferences). 

<img width="684" alt="screen shot 2016-09-16 at 17 54 13" src="https://cloud.githubusercontent.com/assets/1427241/18594194/a6d298ee-7c36-11e6-96cd-c4b4be32111a.png">

I'm confirming with some proper Mac users to see if that would annoy users. 

I'm guessing it's fine. Once this is merged, I'll add a note about it to the README along with one about the leftover file if the user deletes the app.

adam-lynch commented:
OK I got great information from @chillok. We think it's not bad that it doesn't appear in Login Items (even though it's not *nice*) but he let me know that this definitely isn't Mac App Store friendly. An app using this would be rejected because it reaches outside of the app sandbox. See https://developer.apple.com/library/content/documentation/Security/Conceptual/AppSandboxDesignGuide/AboutAppSandbox/AboutAppSandbox.html for more info.

![About sandboxing](https://developer.apple.com/library/content/documentation/Security/Conceptual/AppSandboxDesignGuide/Art/about_sandboxing.png)

![App sandbox directories](https://cloud.githubusercontent.com/assets/1427241/18595375/564c428e-7c3c-11e6-8e84-81119e0eeb52.png)

At first, I thought this was a deal-breaker for this PR but we think the existing method (executing an applescript command) isn't MAS friendly either so we'll go with this PR.

I guess that's just something else I'll have to add to our README. 

---

I know someone will end up coming to this comment in the future after googling so I'll say it is possible to have an auto-launching app in the Mac App Store. See http://martiancraft.com/blog/2015/01/login-items/. As long as it doesn't auto-launch by default, the user can toggle it, etc. You need to have a separate app/executable for this so since that means native code and that extra app/executable would have to be signed just like your main app would, this module would never be able to handle that. I'll create a feature request over on [electron-builder](https://github.com/electron-userland/electron-builder) (as it already handles the signing of Electron apps for you) but I'm not 100% sure it's possible because a user preference is needed. NW.js and Electron might have to add this.

adam-lynch commented:
Here it is: https://github.com/electron-userland/electron-builder/issues/756

adam-lynch commented:
OK after some more overthinking... I'm going to make this an option. So I've merged this as is and I'll sort it out.

adam-lynch commented:
OK I've overhauled this module a bit. There's now a `mac.useLaunchAgent` option. See https://github.com/Teamwork/node-auto-launch/releases/tag/5.0.0. Thanks for the PR @noamokman 👍

noamokman commented:
@adam-lynch thanks!
